test/logging/sink.js

```diff
@@ -0,0 +1,200 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+var assert = require('assert');
+var extend = require('extend');
+var mockery = require('mockery');
+var nodeutil = require('util');
+
+var ServiceObject = require('../../lib/common/service-object.js');
+var util = require('../../lib/common/util.js');
+
+function FakeServiceObject() {
+  this.calledWith_ = arguments;
+  ServiceObject.apply(this, arguments);
+}
+
+nodeutil.inherits(FakeServiceObject, ServiceObject);
+
+describe('Sink', function() {
+  var Sink;
+  var sink;
+
+  var LOGGING = {
+    createSink: util.noop
+  };
+  var SINK_NAME = 'sink-name';
+
+  before(function() {
+    mockery.registerMock('../common/service-object.js', FakeServiceObject);
+
+    mockery.enable({
+      useCleanCache: true,
+      warnOnUnregistered: false
+    });
+
+    Sink = require('../../lib/logging/sink.js');
+  });
+
+  after(function() {
+    mockery.deregisterAll();
+    mockery.disable();
+  });
+
+  beforeEach(function() {
+    sink = new Sink(LOGGING, SINK_NAME);
+  });
+
+  describe('instantiation', function() {
+    it('should inherit from ServiceObject', function(done) {
+      var loggingInstance = extend({}, LOGGING, {
+        createSink: {
+          bind: function(context) {
+            assert.strictEqual(context, loggingInstance);
+            done();
+          }
+        }
+      });
+
+      var sink = new Sink(loggingInstance, SINK_NAME);
+      assert(sink instanceof ServiceObject);
+
+      var calledWith = sink.calledWith_[0];
+
+      assert.strictEqual(calledWith.parent, loggingInstance);
+      assert.strictEqual(calledWith.baseUrl, '/sinks');
+      assert.strictEqual(calledWith.id, SINK_NAME);
+      assert.deepEqual(calledWith.methods, {
+        create: true,
+        delete: true,
+        getMetadata: true
+      });
+    });
+
+    it('should localize the name', function() {
+      assert.strictEqual(sink.name, SINK_NAME);
+    });
+  });
+
+  describe('setFilter', function() {
+    var FILTER = 'filter';
+
+    it('should call set metadata', function(done) {
+      sink.setMetadata = function(metadata, callback) {
+        assert.strictEqual(metadata.filter, FILTER);
+        callback(); // done()
+      };
+
+      sink.setFilter(FILTER, done);
+    });
+  });
+
+  describe('setMetadata', function() {
+    var METADATA = { a: 'b', c: 'd' };
+
+    it('should refresh the metadata', function(done) {
```

Comment:
Because the update (https://cloud.google.com/logging/docs/api/ref/rest/v1beta3/projects.sinks/update) is a PUT vs a PATCH, our API combines the current state of metadata with the input to `setMetadata`, combining them, effectively re-creating PATCH semantics.

Refresh means we get the latest state of the metadata before making the update. This test makes sure we do that.